Microsoft.Identity.Web/TokenAcquisitionB2C.cs

```diff
@@ -0,0 +1,396 @@
+        /// <param name="userTokenCacheProvider">The User token cache provider</param>
+        public TokenAcquisitionB2C(
+            IMsalAppTokenCacheProvider appTokenCacheProvider,
+            IMsalUserTokenCacheProvider userTokenCacheProvider,
+            IHttpContextAccessor httpContextAccessor,
+            IOptions<AzureADB2COptions> azureAdOptions,
+            IOptions<ConfidentialClientApplicationOptions> applicationOptions)
+        {
+            _httpContextAccessor = httpContextAccessor;
+            _azureAdB2COptions = azureAdOptions.Value;
+            _applicationOptions = applicationOptions.Value;
+            _appTokenCacheProvider = appTokenCacheProvider;
+            _userTokenCacheProvider = userTokenCacheProvider;
+        }
+
+        /// <summary>
+        /// Scopes which are already requested by MSAL.NET. they should not be re-requested;
+        /// </summary>
+        private readonly string[] _scopesRequestedByMsalNet = new string[]
+        {
+            OidcConstants.ScopeOpenId,
+            OidcConstants.ScopeProfile,
+            OidcConstants.ScopeOfflineAccess
+        };
+
+        public async Task AddAccountToCacheFromAuthorizationCodeAsync(AuthorizationCodeReceivedContext context, IEnumerable<string> scopes)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (scopes == null)
+            {
+                throw new ArgumentNullException(nameof(scopes));
+            }
+
+            try
+            {
+                context.HandleCodeRedemption();
+
+                if (!context.HttpContext.User.Claims.Any())
+                {
+                    (context.HttpContext.User.Identity as ClaimsIdentity).AddClaims(context.Principal.Claims);
+                }
+
+                var application = GetOrBuildConfidentialClientApplication();
+
+                var result = await application
+                    .AcquireTokenByAuthorizationCode(scopes.Except(_scopesRequestedByMsalNet), context.ProtocolMessage.Code)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                context.HandleCodeRedemption(null, result.IdToken);
+            }
+            catch (MsalException ex)
+            {
+                // brentsch - todo, write to a log
+                Debug.WriteLine(ex.Message);
+                throw;
+            }
+        }
+
+        public async Task<string> GetAccessTokenOnBehalfOfUserAsync(
+            IEnumerable<string> scopes,
+            string tenant = null)
+        {
+            if (scopes == null)
+            {
+                throw new ArgumentNullException(nameof(scopes));
+            }
+
+            // Use MSAL to get the right token to call the API
+            var application = GetOrBuildConfidentialClientApplication();
+            
+            // Case of a lazy OBO
```

Comment:
>// Case of a lazy OBO [](start = 12, length = 21)

idk why we still have this? i thought this was removed. this gets super confusing because it's not really OBO as OBO is intended to be used. but this can be discussed offline

Reply:
I am with you on this one @jennyf19. I honestly dont like the name OBO for this case because it is not necessarily using the OBO token acquisition flow. I copied this from the `TokenAcquisition` class and decided not to touch since this PR had another purpose. On the other hand, I understand @jmprieur s idea to have a single method that is responsible for getting an access token, simplifying the developers side.

I see this as deeper discussion that we could have later.

Reply:
yes...i think this has been changed w/the earlier PR and @jmprieur  is in the process of fixing this. so probably a non-issue, but let's keep it on our radar for now. :)